cmd/runtimetest/main.go

```diff
@@ -410,201 +578,379 @@ func validateDefaultFS(spec *rspec.Spec, t *tap.T) error {
+			Path: "/dev/console",
+			Type: "c",
+			// FIXME: get the major/minor from the controlling TTY
+		})
 	}
 
 	for _, device := range defaultDevices {
-		fi, err := os.Stat(device)
+		err := c.validateDevice(
+			&device,
+			specerror.DefaultDevices,
+			spec.Version,
+			fmt.Sprintf("%s (default device)", device.Path))
 		if err != nil {
-			if os.IsNotExist(err) {
-				return specerror.NewError(specerror.DefaultDevices,
-					fmt.Errorf("device node %v not found", device),
-					rspec.Version)
-			}
 			return err
 		}
-		if fi.Mode()&os.ModeDevice != os.ModeDevice {
-			return specerror.NewError(specerror.DefaultDevices,
-				fmt.Errorf("file %v is not a device as expected", device),
-				rspec.Version)
-		}
 	}
 
 	return nil
 }
 
-func validateMaskedPaths(spec *rspec.Spec, t *tap.T) error {
-	if spec.Linux == nil {
+func (c *complianceTester) validateMaskedPaths(spec *rspec.Spec) error {
+	if spec.Linux == nil || spec.Linux.MaskedPaths == nil {
+		c.harness.Skip(1, "linux.maskedPaths not set")
 		return nil
 	}
+
 	for _, maskedPath := range spec.Linux.MaskedPaths {
-		f, err := os.Open(maskedPath)
-		if err != nil {
+		readable, err := testReadAccess(maskedPath)
+		if err != nil && !os.IsNotExist(err) {
 			return err
 		}
-		defer f.Close()
-		b := make([]byte, 1)
-		_, err = f.Read(b)
-		if err != io.EOF {
-			return fmt.Errorf("%v should not be readable", maskedPath)
-		}
+		c.harness.Ok(!readable, fmt.Sprintf("cannot read masked path %q", maskedPath))
 	}
+
 	return nil
 }
 
-func validateSeccomp(spec *rspec.Spec, t *tap.T) error {
+func (c *complianceTester) validateSeccomp(spec *rspec.Spec) error {
 	if spec.Linux == nil || spec.Linux.Seccomp == nil {
+		c.harness.Skip(1, "linux.seccomp not set")
 		return nil
 	}
+
 	for _, sys := range spec.Linux.Seccomp.Syscalls {
 		if sys.Action == "SCMP_ACT_ERRNO" {
 			for _, name := range sys.Names {
 				if name == "getcwd" {
 					_, err := os.Getwd()
 					if err == nil {
-						t.Diagnostic("getcwd did not return an error")
+						c.harness.Skip(1, "getcwd did not return an error")
 					}
 				} else {
-					t.Skip(1, fmt.Sprintf("%s syscall returns errno", name))
+					c.harness.Skip(1, fmt.Sprintf("%s syscall returns errno", name))
 				}
 			}
 		} else {
-			t.Skip(1, fmt.Sprintf("syscall action %s", sys.Action))
+			c.harness.Skip(1, fmt.Sprintf("syscall action %s", sys.Action))
 		}
 	}
+
 	return nil
 }
 
-func validateROPaths(spec *rspec.Spec, t *tap.T) error {
-	if spec.Linux == nil {
+func (c *complianceTester) validateROPaths(spec *rspec.Spec) error {
+	if spec.Linux == nil || spec.Linux.ReadonlyPaths == nil {
+		c.harness.Skip(1, "linux.readonlyPaths not set")
 		return nil
 	}
-	for _, v := range spec.Linux.ReadonlyPaths {
-		err := testWriteAccess(v)
-		if err == nil {
-			return fmt.Errorf("%v should be readonly", v)
+
+	for i, path := range spec.Linux.ReadonlyPaths {
+		readable, err := testReadAccess(path)
+		if err != nil && !os.IsNotExist(err) {
```

Comment:
Why ignoring the `os.IsNotExist` error?

Reply:
> Why ignoring the `os.IsNotExist` error?

Because if the path is not there at all, it's certainly not readable ;).  But while that makes some sense for `maskedPaths`, runtimes should probably not go that far with `readonlyPaths`.  I've dropped it from the `readonlyPaths` checker in 031e4c8 → 5e395e8.